eclipse-theia/theia
# Support find widget in notebooks

#### What it does

Adds support for a monaco-like find widget to the notebook widget. 

#### How to test

The search widget can be brought up by pressing `Ctrl+F` inside of any area of the notebook widget. Having the Jupyter Keymaps extension installed also allows to bring it up by just pressing `F` outside of an editor area (while the notebook widget is focused).

Multiple test scenarios:
* In general: Test that the widget behaves similarly to the existing find widget for text editors.
* Select text in a monaco editor and press `Ctrl+F`. The selected text should be placed as the initial search value of the find widget.
* Search for text in code. The code should be highlighted (with the current selection getting a special highlighting) and jumping to different matches should work.
* Search using match case, word boundaries and regex should work as expected.
* Search for text in rendered markdown cells. The text should be highlighted directly in the HTML. Jumping should work as well.
* Replacing code in the code editors should work as expected.
* Replacing code in the rendered markdown cells should not be possible (button greyed out).
* Upon changing the editing mode of a markdown cell, the search should update and highlight the now rendered code editor instead (also in reverse).

#### Follow-ups

Known issues:
* Cannot find in webviews
* Cannot replace on rendered markdown
* There is no global undo for the global replace yet. Each cell change needs to be undone individually.
* Performance for huge rendered markdown is quite bad

The current API is pretty suboptimal (especially for finding matches inside of rendered markdown cells). I plan to get back to this later to revise the API and extract a common `FindWidget` that can be used by any widget.

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
